File: Data Structures & Algorithms/duplicate-integer/submission-0.py
class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        num_in_nums = {}
        is_duplicate = False

        for i in nums:
            if num_in_nums.get(i) == None:
                num_in_nums[i] = 1
            else:
                is_duplicate = True
                break
        
        return is_duplicate

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int nums_size = nums.size();
        vector<int> pref(nums_size), suff(nums_size), res(nums_size);
        pref[0] = 1;
        suff[nums.size() - 1] = 1;

        int foo = 1;
        for (int i = 1; i < nums_size; i++) {
            foo *= nums[i-1];
            pref[i] = foo;
        }

        foo = 1;
        for (int i = nums_size - 2; i >= 0; i--) {
            foo *= nums[i+1];
            suff[i] = foo;
        }

        for (int i = 0; i < nums_size; i++) {
            res[i] = pref[i] * suff[i];
        }
        return res;
    }
};
